qhub/template/{{ cookiecutter.repo_directory }}/.gitlab-ci.yml

```diff
@@ -0,0 +1,96 @@
+services:
+  - docker:dind
+
+render-qhub:
+  image: "python:3.9"
+  variables:
+    COMMIT_MSG: |
+      qhub-config.yaml automated commit: {{ '$CI_COMMIT_SHA' }}
+  script:
+    - pip install qhub==0.2.3
+    - qhub render -c qhub-config.yaml
+    - git config user.email "qhub@quansight.com"
+    - git config user.name "gitlab ci"
+    - git add .
+    - git diff --quiet && git diff --staged --quiet || (git commit -m "${COMMIT_MSG}"; git push origin master)
+  only:
+    refs:
+      - master
+    changes:
+      - "qhub-config.yaml"
+
+build-docker-images:
+  image: "docker:19.03.12"
+  timeout: 2h
+  parallel:
+    matrix:
+      - DOCKERFILE:
+          - jupyterlab
+          - jupyterhub
+          - dask-worker
+  variables:
+    IMAGE_TAG: {{ '$CI_COMMIT_SHA' }}
+    IMAGE_NAME: {{ cookiecutter.project_name }}-{{ '$DOCKERFILE' }}
+  script:
+    - docker build -f "image/Dockerfile.{{ '$DOCKERFILE' }}" -t "$IMAGE_NAME:$IMAGE_TAG" image
+  only:
+    changes:
+      - "image/**"
+
+terraform-jupyterhub:
+  image:
+    name: "hashicorp/terraform"
+    entrypoint: [""]
+  variables: 
+    PROVIDER: ${ cookiecutter.provider }
+    PREFECT: ${ cookiecutter.prefect }   
+    WORKSPACE: infrastructure/
+    VARS: ""
+  rules:
+    - if: $CI_COMMIT_REF_NAME == /master/
+      changes:
+        - "infrastructure/**" 
+        - "environments/**"
+    
+    - if: $PREFECT
+      variables:
+        VARS: '-var="prefect_token=${{ secrets.PREFECT_TOKEN }}"'
+
+    - if: $PROVIDER == 'aws'
+      variables:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

Comment:
I will change this `secrets.GITHUB_TOKEN` later, we need to match this with Gitlab keys right?